File: src/localmax_runner/canonical.py
"""RFC 8785 JSON Canonicalization Scheme.

The manifest is signed over its canonical form, and the API verifies that signature with an
independent implementation. Both sides must agree byte for byte, so key ordering, whitespace
and number formatting are all fixed here rather than left to the JSON encoder.
"""

from __future__ import annotations

import hashlib
import json
import math
from typing import Any


def _number(value: float | int) -> str:
    if isinstance(value, bool):  # bool is a subclass of int; must not reach here
        raise TypeError("bool is not a number")
    if isinstance(value, int):
        return str(value)
    if not math.isfinite(value):
        raise ValueError("Cannot canonicalize a non-finite number")
    if value == int(value) and abs(value) < 1e21:
        # ES6 Number#toString renders integral floats without a fractional part, and the
        # JavaScript verifier on the other side will do exactly that.
        return str(int(value))
    return repr(value)


def canonicalize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return _number(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonicalize(v) for v in value) + "]"
    if isinstance(value, dict):
        items = sorted((k for k in value if value[k] is not None), key=lambda k: k.encode("utf-16-be"))
        return (
            "{"
            + ",".join(
                f"{json.dumps(k, ensure_ascii=False, separators=(',', ':'))}:{canonicalize(value[k])}"
                for k in items
            )
            + "}"
        )
    raise TypeError(f"Cannot canonicalize {type(value).__name__}")


def signing_payload(manifest: dict[str, Any]) -> bytes:
    """The exact bytes signed: the manifest with `signature.value` removed."""
    copy = dict(manifest)
    signature = copy.get("signature")
    if isinstance(signature, dict):
        copy["signature"] = {k: v for k, v in signature.items() if k != "value"}
    return canonicalize(copy).encode("utf-8")


def sha256_file(path, chunk: int = 1024 * 1024) -> tuple[str, int]:
    """Return (prefixed hex digest, byte size) for a file, streaming it."""
    digest = hashlib.sha256()
    size = 0
    with open(path, "rb") as handle:
        while True:
            block = handle.read(chunk)
            if not block:
                break
            digest.update(block)
            size += len(block)
    return f"sha256:{digest.hexdigest()}", size


def sha256_text(text: str) -> str:
    return "sha256:" + hashlib.sha256(text.encode("utf-8")).hexdigest()

File: apps/api/src/lib/canonical.ts
/**
 * RFC 8785 JSON Canonicalization Scheme.
 *
 * The runner signs the canonical form of the manifest with the `signature.value` field
 * removed, so both sides must agree byte for byte on the serialization. Key insertion
 * order, whitespace and number formatting must not be able to change the bytes.
 */

/** ES6 number-to-string, which RFC 8785 mandates for numeric serialization. */
function serializeNumber(n: number): string {
  if (!Number.isFinite(n)) {
    throw new TypeError("Cannot canonicalize a non-finite number");
  }
  if (Object.is(n, -0)) return "0";
  return String(n);
}

function serializeString(s: string): string {
  // JSON.stringify already emits the escaping RFC 8785 requires (\b \t \n \f \r \" \\ and
  // \u00xx for the remaining control characters), and leaves all other code points literal.
  return JSON.stringify(s);
}

export function canonicalize(value: unknown): string {
  if (value === null) return "null";

  switch (typeof value) {
    case "boolean":
      return value ? "true" : "false";
    case "number":
      return serializeNumber(value);
    case "string":
      return serializeString(value);
    case "object":
      break;
    default:
      throw new TypeError(`Cannot canonicalize ${typeof value}`);
  }

  if (Array.isArray(value)) {
    return `[${value.map(canonicalize).join(",")}]`;
  }

  const obj = value as Record<string, unknown>;
  // RFC 8785 sorts by UTF-16 code unit, which is what Array#sort does by default.
  const keys = Object.keys(obj).filter((k) => obj[k] !== undefined).sort();
  const parts = keys.map((k) => `${serializeString(k)}:${canonicalize(obj[k])}`);
  return `{${parts.join(",")}}`;
}

/** The exact bytes the runner signed: the manifest without `signature.value`. */
export function signingPayload(manifest: Record<string, unknown>): string {
  const copy: Record<string, unknown> = { ...manifest };
  const sig = copy.signature as Record<string, unknown> | undefined;
  if (sig) {
    const { value: _omit, ...rest } = sig;
    copy.signature = rest;
  }
  return canonicalize(copy);
}
